cmd/query/app/querysvc/adjuster/hash.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package adjuster
+
+import (
+	"hash"
+	"hash/fnv"
+
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+var _ Adjuster = (*SpanHashDeduper)(nil)
+
+// SpanHash creates an adjuster that deduplicates spans by removing all but one span
+// with the same hash code. This is particularly useful for scenarios where spans
+// may be duplicated during archival, such as with ElasticSearch archival.
+//
+// The hash code is generated by serializing the span into protobuf bytes and applying
+// the FNV hashing algorithm to the serialized data.
+//
+// To ensure consistent hash codes, this adjuster should be executed after
+// SortAttributesAndEvents, which normalizes the order of collections within the span.
+func SpanHash() SpanHashDeduper {
+	return SpanHashDeduper{
+		marshaller: &ptrace.ProtoMarshaler{},
+		hasher:     fnv.New64a(),
+	}
+}
+
+type SpanHashDeduper struct {
+	marshaller ptrace.Marshaler
+	hasher     hash.Hash64
+}
+
+func (s *SpanHashDeduper) Adjust(traces ptrace.Traces) {
+	resourceSpans := traces.ResourceSpans()
+	for i := 0; i < resourceSpans.Len(); i++ {
+		rs := resourceSpans.At(i)
+		scopeSpans := rs.ScopeSpans()
+		for j := 0; j < scopeSpans.Len(); j++ {
+			ss := scopeSpans.At(j)
+			spansByHash := make(map[uint64]ptrace.Span)
+			spans := ss.Spans()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+				hash, err := s.computeHashCode(span)
+				if err != nil {
+					// TODO: what should we do here?
+					continue
+				}
```

Comment:
@yurishkuro how should we handle the case where the hash code cannot be computed. This would happen in case there as an error in protobuf serialization or if the hashing function returned an error. Its probably very unlikely this ever happens. Is skipping over the span sufficient? Do we want to add a warning?

Reply:
yeah, I think skipping the span is fine in this case. We could also add a warning with the error to that span.